File: internal/handlers/book.go
package handlers

import (
	"net/http"

	"github.com/chumnend/pook/internal/models"
	"github.com/chumnend/pook/internal/utils"
	"github.com/google/uuid"
)

func CreateBook(w http.ResponseWriter, req *http.Request) {
	type requestInput struct {
		UserID string `json:"userId"`
		Title  string `json:"title"`
	}

	var input requestInput
	if err := utils.ParseJSON(req, &input); err != nil {
		http.Error(w, "invalid input", http.StatusBadRequest)
		return
	}

	if input.Title == "" || input.UserID == "" {
		http.Error(w, "all fields (title, userId) are required", http.StatusBadRequest)
		return
	}

	parsedUUID, err := uuid.Parse(input.UserID)
	if err != nil {
		http.Error(w, "invalid user id", http.StatusBadRequest)
		return
	}

	if err := models.CreateBook(parsedUUID, input.Title); err != nil {
		http.Error(w, "unable to create book", http.StatusBadRequest)
		return
	}

	response := map[string]string{
		"message": "book successfully created",
	}

	utils.SendJSON(w, response, http.StatusOK)
}

func GetAllBooks(w http.ResponseWriter, req *http.Request) {
	userID := req.URL.Query().Get("user_id")
	if userID != "" {
		parsedUUID, err := uuid.Parse(userID)
		if err != nil {
			http.Error(w, "invalid user id", http.StatusBadRequest)
			return
		}

		books, err := models.GetBooksByUserID(parsedUUID)
		if err != nil {
			http.Error(w, "unable to get user's books", http.StatusInternalServerError)
			return
		}

		response := map[string]any{
			"books": books,
		}

		utils.SendJSON(w, response, http.StatusOK)
	}

	books, err := models.GetAllBooks()
	if err != nil {
		http.Error(w, "unable to get all book", http.StatusInternalServerError)
		return
	}

	response := map[string]any{
		"books": books,
	}

	utils.SendJSON(w, response, http.StatusOK)
}

func GetBook(w http.ResponseWriter, req *http.Request) {
	bookID := req.PathValue("book_id")

	parsedUUID, err := uuid.Parse(bookID)
	if err != nil {
		http.Error(w, "invalid book_id", http.StatusBadRequest)
		return
	}

	book, err := models.GetBookByID(parsedUUID)
	if err != nil {
		http.Error(w, "book not found", http.StatusNotFound)
		return
	}

	response := map[string]any{
		"book": book,
	}

	utils.SendJSON(w, response, http.StatusOK)
}

func UpdateBook(w http.ResponseWriter, req *http.Request) {
	bookID := req.PathValue("book_id")

	parsedUUID, err := uuid.Parse(bookID)
	if err != nil {
		http.Error(w, "invalid book_id", http.StatusBadRequest)
		return
	}

	type requestInput struct {
		Title string `json:"title"`
	}

	var input requestInput
	if err := utils.ParseJSON(req, &input); err != nil {
		http.Error(w, "invalid input", http.StatusBadRequest)
		return
	}

	if input.Title == "" {
		http.Error(w, "title is required", http.StatusBadRequest)
		return
	}

	if err := models.UpdateBookByID(parsedUUID, input.Title); err != nil {
		http.Error(w, "unable to update book", http.StatusInternalServerError)
		return
	}

	response := map[string]string{
		"message": "book successfully updated",
	}

	utils.SendJSON(w, response, http.StatusOK)
}

func DeleteBook(w http.ResponseWriter, req *http.Request) {
	bookID := req.PathValue("book_id")

	parsedUUID, err := uuid.Parse(bookID)
	if err != nil {
		http.Error(w, "invalid book_id", http.StatusBadRequest)
		return
	}

	if err := models.DeleteBookByID(parsedUUID); err != nil {
		http.Error(w, "unable to delete book", http.StatusInternalServerError)
		return
	}

	response := map[string]string{
		"message": "book successfully deleted",
	}

	utils.SendJSON(w, response, http.StatusOK)
}

File: web/src/helpers/services/book.ts
const API_BASE_URL = import.meta.env.VITE_API_BASE_URL;

export const createBook = async (userId: string, title: string) => {
  const response = await fetch(`${API_BASE_URL}/books`, {
    method: 'POST',
    headers: {
      'Content-Type': 'application/json',
    },
    body: JSON.stringify({ userId, title }),
  });

  if (!response.ok) {
    const error = await response.text();
    throw new Error(error || `HTTP Error: ${response.status}`);
  }

  return response.json();
};

export const getAllBooks = async() => {
  const response = await fetch(`${API_BASE_URL}/books` , {
    method: 'GET',
    headers: {
      'Content-Type': 'application/json',
    },
  });

  if (!response.ok) {
    const error = await response.text();
    throw new Error(error || `HTTP Error: ${response.status}`);
  }

  return response.json();
}

export const getAllBooksByUserId = async (userId: string) => {
  const response = await fetch(`${API_BASE_URL}/books?user_id=${encodeURIComponent(userId)}`, {
    method: 'GET',
    headers: {
      'Content-Type': 'application/json',
    },
  });

  if (!response.ok) {
    const error = await response.text();
    throw new Error(error || `HTTP Error: ${response.status}`);
  }

  return response.json();
};

export const getBookById = async (bookId: string) => {
  const response = await fetch(`${API_BASE_URL}/books/${encodeURIComponent(bookId)}`, {
    method: 'GET',
    headers: {
      'Content-Type': 'application/json',
    },
  });

  if (!response.ok) {
    const error = await response.text();
    throw new Error(error || `HTTP Error: ${response.status}`);
  }

  return response.json();
};

export const updateBook = async (bookId: string, title: string) => {
  const response = await fetch(`${API_BASE_URL}/books/${encodeURIComponent(bookId)}`, {
    method: 'PUT',
    headers: {
      'Content-Type': 'application/json',
    },
    body: JSON.stringify({ title }),
  });

  if (!response.ok) {
    const error = await response.text();
    throw new Error(error || `HTTP Error: ${response.status}`);
  }

  return response.json();
};

export const deleteBook = async (bookId: string) => {
  const response = await fetch(`${API_BASE_URL}/books/${encodeURIComponent(bookId)}`, {
    method: 'DELETE',
    headers: {
      'Content-Type': 'application/json',
    },
  });

  if (!response.ok) {
    const error = await response.text();
    throw new Error(error || `HTTP Error: ${response.status}`);
  }

  return response.json();
};
